kube-client/src/client/auth/mod.rs

```diff
@@ -1,18 +1,75 @@
-use std::{process::Command, sync::Arc};
+use std::{path::PathBuf, process::Command, sync::Arc};
 
 use chrono::{DateTime, Duration, Utc};
-use http::HeaderValue;
+use http::{header::InvalidHeaderValue, HeaderValue};
 use jsonpath_lib::select as jsonpath_select;
 use serde::{Deserialize, Serialize};
+use thiserror::Error;
 use tokio::sync::Mutex;
 
-use crate::{
-    config::{read_file_to_string, AuthInfo, AuthProviderConfig, ExecConfig},
-    error::{ConfigError, Error},
-    Result,
-};
+use crate::config::{AuthInfo, AuthProviderConfig, ExecConfig};
 
 #[cfg(feature = "oauth")] mod oauth;
+#[cfg(feature = "oauth")] pub use oauth::Error as OAuthError;
+
+#[derive(Error, Debug)]
+/// Client auth errors
+pub enum Error {
+    /// Invalid basic auth
+    #[error("invalid basic auth: {0}")]
+    InvalidBasicAuth(#[source] InvalidHeaderValue),
+
+    /// Invalid bearer token
+    #[error("invalid bearer token: {0}")]
+    InvalidBearerToken(#[source] InvalidHeaderValue),
+
+    /// Tried to refresh a token and got a non-refreshable token response
+    #[error("tried to refresh a token and got a non-refreshable token response")]
+    UnrefreshableTokenResponse,
+
+    /// Exec plugin response did not contain a status
+    #[error("exec-plugin response did not contain a status")]
+    ExecPluginFailed,
+
+    /// Malformed token expiration date
+    #[error("malformed token expiration date: {0}")]
+    MalformedTokenExpirationDate(#[source] chrono::ParseError),
+
+    /// Failed to start auth exec
+    #[error("unable to run auth exec: {0}")]
+    AuthExecStart(#[source] std::io::Error),
+
+    /// Failed to run auth exec command
+    #[error("auth exec command '{cmd}' failed with status {status}: {out:?}")]
+    #[allow(missing_docs)]
+    AuthExecRun {
+        cmd: String,
+        status: std::process::ExitStatus,
+        out: std::process::Output,
+    },
+
+    /// Failed to parse auth exec output
+    #[error("failed to parse auth exec output: {0}")]
+    AuthExecParse(#[source] serde_json::Error),
+
+    /// Failed to exec auth
+    #[error("failed exec auth: {0}")]
+    AuthExec(String),
```

Comment:
`AuthExec(String)` should be broken down, but we can do that later.
In general, we shouldn't have code like `Error::Foo(format!("result is not a string {:?}", e))`.